packages/rum-core/src/domain/contextHistory.ts

```diff
@@ -0,0 +1,66 @@
+import { ONE_MINUTE, ClocksState, RelativeTime, relativeNow } from '@datadog/browser-core'
+
+interface PreviousContext<T> {
+  startTime: RelativeTime
+  endTime: RelativeTime
+  context: T
+}
+
+export const CLEAR_OLD_CONTEXTS_INTERVAL = ONE_MINUTE
+
+export class ContextHistory<Raw extends { startClocks: ClocksState }, Built> {
```

Comment:
It seems like only `startClocks.relative` is used, maybe change Raw to extend `{ startTime: RelativeTime }` ?

Reply:
Indeed, internally we are only using relative time, but for parent contexts usage we'd like to use it with objects containing clocks (ViewCreatedEvent, AutoActionCreatedEvent).
Adding a startTime to these objects or creating new ones with it does not seem so great to me, converting everything to clock is a bit painful as well.
We could not force the `Raw` object structure and provide the current context start time outside of the current context object, it would allow us to work only with relative time.

something like:
```ts
const foo = { value, startClocks }
contextHistory.current = foo
contextHistory.currentStartTime = foo.startClocks.relative
```

wdyt?

Reply:
What is the benefit of a class instead of a function?

Reply:
What is the benefit of a class instead of a function?

Reply:
> What is the benefit of a class instead of a function?

when I need to encapsulate some logic for a behavior that I want to use independently at several places, I tend to use classes by default. 
We could achieve a similar result with functions.

Reply:
Even though it's just code style, I would move to a functional implementation for consistency reason then.

Reply:
there are several other examples of classes in the code base.
it is a quite common concept, any issue with it?

Reply:
> Indeed, internally we are only using relative time, but for parent contexts usage we'd like to use it with objects containing clocks (ViewCreatedEvent, AutoActionCreatedEvent).
> Adding a startTime to these objects or creating new ones with it does not seem so great to me, converting everything to clock is a bit painful as well.
> We could not force the `Raw` object structure and provide the current context start time outside of the current context object, it would allow us to work only with relative time.
> 
> something like:
> 
> ```ts
> const foo = { value, startClocks }
> contextHistory.current = foo
> contextHistory.currentStartTime = foo.startClocks.relative
> ```
> 
> wdyt?

It looks good, and maybe we could go one step further and remove the `buildContext` function altogether. For the view context case, we could have:

```ts
contextHistory.current = {
      session: {
        id: viewCreatedEvent.sessionId,
      },
      view: {
        id: viewCreatedEvent.id,
        name: viewCreatedEvent.name,
        referrer: viewCreatedEvent.referrer,
        url: viewCreatedEvent.location.href,
      },
    }
context.currentStartTime = viewCreatedEvent.startClocks.relative
```

Maybe an `updateCurrent(current, startTime)` method would be nicer.

Reply:
> Maybe an updateCurrent(current, startTime) method would be nicer.

👍 I started an implementation with something like that

> maybe we could go one step further and remove the `buildContext` function altogether

ok so the client is responsible to keep the info needed to generate the context.
it should remove the issue with https://github.com/DataDog/browser-sdk/pull/1057#discussion_r710937948 then 😉

Reply:
In fact, [this test](https://github.com/DataDog/browser-sdk/blob/main/packages/rum-core/src/domain/parentContexts.spec.ts#L121) is preventing us to remove the `buildContext`  for now, but we should be able to remove it in the last PR.